src/radix_tree.cpp

```diff
@@ -0,0 +1,602 @@
+    // deleting the last byte in the region of first bytes by
+    // overwriting it.
+    memmove(parent_node.node_ptrs() - 1,
+                 parent_node.node_ptrs(),
+                 parent_node.edgecount() * sizeof(void*));
+
+    // Shrink the parent node to the new size, which "deletes" the
+    // last pointer in the chunk of node pointers.
+    parent_node.resize(parent_node.prefix_length(),
+                       parent_node.edgecount() - 1);
+
+    // Nothing points to this node now, so we can reclaim it.
+    free(current_node.data_);
+
+    if (parent_node.prefix_length() == 0)
+        root_.data_ = parent_node.data_;
+    else
+        grandparent_node.set_node_at(gp_edge_idx, parent_node);
+    return true;
+}
+
+bool zmq::radix_tree::check(unsigned char* key, size_t size)
+{
+    if (root_.refcount() > 0)
+        return true;
+
+    match_result result = match(key, size, true);
+    return result.nkey == size
+        && result.nprefix == result.current_node.prefix_length()
+        && result.current_node.refcount() > 0;
+}
+
+static void visit_keys(node n,
+                       unsigned char** buffer,
+                       size_t buffer_size,
+                       size_t maxbuffer_size,
+                       void (*func)(unsigned char* data,
+                                    size_t size,
+                                    void* arg),
+                       void *arg)
+{
+    if (buffer_size >= maxbuffer_size) {
+        maxbuffer_size += 256;
+        *buffer = static_cast<unsigned char*>(realloc(*buffer, maxbuffer_size));
+        zmq_assert(*buffer);
+    }
+
+    for (size_t i = 0; i < n.prefix_length(); ++i)
+        (*buffer)[buffer_size++] = n.prefix()[i];
+    if (n.refcount() > 0)
+        func(*buffer, buffer_size, arg);
+    for (size_t i = 0; i < n.edgecount(); ++i)
+        visit_keys(n.node_at(i), buffer, buffer_size,
+                   maxbuffer_size, func, arg);
+    buffer_size -= n.prefix_length();
+}
+
+void zmq::radix_tree::apply(void (*func)(unsigned char* data,
+                                    size_t size,
+                                    void* arg),
+                       void* arg)
+{
+    unsigned char* buffer = NULL;
+    visit_keys(root_, &buffer, 0, 0, func, arg);
+    free(buffer);
+}
+
+size_t zmq::radix_tree::size() const
+{
+    return size_;
+}
+
+static void visit_child(node child_node, size_t level)
+{
+    zmq_assert(level > 0);
+
+    for (size_t i = 0; i < 4 * (level - 1) + level; ++i)
+        putchar(' ');
+    printf("`-> ");
+    for (uint32_t i = 0; i < child_node.prefix_length(); ++i)
+        printf("%c", child_node.prefix()[i]);
+    if (child_node.refcount() > 0)
+        printf(" [*]");
+    printf("\n");
+    for (uint32_t i = 0; i < child_node.edgecount(); ++i)
+        visit_child(child_node.node_at(i), level + 1);
+}
+
+void zmq::radix_tree::print()
+{
+    puts("[root]");
+    for (uint32_t i = 0; i < root_.edgecount(); ++i)
```

Comment:
This is probably only for testing purposes. Can this be moved to the tests? The library code should not write to stdout